packages/priority-queue/src/test/index.js

```diff
@@ -0,0 +1,34 @@
+/**
+ * External dependencies
+ */
+import { EventEmitter } from 'events';
+
+/**
+ * Internal dependencies
+ */
+import { createQueue } from '../';
+import requestIdleCallback from '../request-idle-callback';
+
+const emitter = new EventEmitter();
+
+const tick = () => emitter.emit( 'tick' );
+
+jest.mock( '../request-idle-callback', () => jest.fn() );
+
+requestIdleCallback.mockImplementation( ( callback ) => {
+	emitter.once( 'tick', () => callback( Date.now() ) );
+} );
```

Comment:
You can just return the implementation from the mock factory function (second parameter to `jest.mock`).

Reply:
> You can just return the implementation from the mock factory function (second parameter to `jest.mock`).

It was originally implemented this way, but it can't be done, because `jest.mock` does not allow references to the outside scope within the callback:

```
 FAIL  packages/priority-queue/src/test/index.js
  ● Test suite failed to run

    babel-plugin-jest-hoist: The module factory of `jest.mock()` is not allowed to reference any out-of-scope variables.
    Invalid variable access: emitter
```

Reply:
That makes me think that `tick` should be returned from the mock factory instead of being defined in the file.

Reply:
> That makes me think that `tick` should be returned from the mock factory instead of being defined in the file.

How do you imagine that looking? It still needs to satisfy the requirement of being called as the `requestIdleCallback` within the `priority-queue/src/index.js` implementation, so I guess it would have to be some custom property on the function to do the tick?

Closest thing I've been able to get working is this, which I don't personally think is an improvement.

```diff
diff --git a/packages/priority-queue/src/test/index.js b/packages/priority-queue/src/test/index.js
index 57f519d54..19ecb29a8 100644
--- a/packages/priority-queue/src/test/index.js
+++ b/packages/priority-queue/src/test/index.js
@@ -1,22 +1,16 @@
-/**
- * External dependencies
- */
-import { EventEmitter } from 'events';
-
 /**
  * Internal dependencies
  */
 import { createQueue } from '../';
 import requestIdleCallback from '../request-idle-callback';
 
-const emitter = new EventEmitter();
-
-const tick = () => emitter.emit( 'tick' );
-
-jest.mock( '../request-idle-callback', () => jest.fn() );
+jest.mock( '../request-idle-callback', () => {
+	const emitter = new ( require.requireActual( 'events' ).EventEmitter )();
 
-requestIdleCallback.mockImplementation( ( callback ) => {
-	emitter.once( 'tick', () => callback( Date.now() ) );
+	return Object.assign(
+		( callback ) => emitter.once( 'tick', () => callback( Date.now() ) ),
+		{ tick: () => emitter.emit( 'tick' ) },
+	);
 } );
 
 describe( 'createQueue', () => {
@@ -28,7 +22,7 @@ describe( 'createQueue', () => {
 	it( 'runs callback after processing waiting queue', () => {
 		const callback = jest.fn();
 		queue.add( {}, callback );
-		tick();
+		requestIdleCallback.tick();
 		expect( callback ).toHaveBeenCalled();
 	} );
 } );
```

Reply:
Yes, like that. That way the mock could be shared between test files in the future.

Reply:
> Yes, like that. That way the mock could be shared between test files in the future.

Updated in 338ea7255.